### PR TITLE
client/dcr: Mixing Tx History

### LINKED ISSUE #2824: Privacy Mixing Transactions labeled "Send" - Scary UX
image

Novice users might think their wallet is being drained with all the "Send" Txs and red "amount".

Suggestion: Change "Send" to "Mix" or "Privacy" to avoid scares.

Change the red numbers in the Amount section to a different color like yellow.

### PR DESCRIPTION
**Pull request by martonp**
This diff handles identifying decred mixing transactions when adding them to the transaction history. For RPC wallets, a separate transaction history db is created for each set of unmixed/trading/mixed accounts configuration.

Closes #2824

### CODE REVIEW

**Review by JoeGruffins — APPROVED** [review]

File: client/asset/dcr/spv.go [hunk 1]
```diff
@@ -549,6 +548,134 @@ func (w *spvWallet) AddressInfo(ctx context.Context, addrStr string) (*AddressIn
...
+		}
+
+		inMsgTx, err := msgTxFromHex(inTx.Hex)
+		if err != nil {
+			return false, 0, 0, err
+		}
+
+		if len(inMsgTx.TxOut) <= int(txIn.PreviousOutPoint.Index) {
+			continue
+		}
```
[the remark this answers is not in the document]
Comment (martonp) [reply client/asset/dcr/spv.go:610]: Yeah, but either we panic, ignore the whole transaction, or do this. This shouldn't happen unless there's a big problem somewhere else.

[the remark this answers is not in the document]
Comment (buck54321) [reply client/asset/dcr/spv.go:610]: Let's return an error.

File: client/asset/dcr/dcr.go [hunk 2]
```diff
@@ -935,33 +935,101 @@ func (dcr *ExchangeWallet) findExistingAddressBasedTxHistoryDB() (string, error)
...
+func (dcr *ExchangeWallet) findRPCTxHistoryDBFile(ctx context.Context, wallet *rpcWallet) (string, error) {
+	dbInfos, err := dcr.existingRPCTxHistoryDBs()
+	if err != nil {
+		return "", err
 	}
 
-	if dbPath == "" {
-		addressPath, err := dcr.findExistingAddressBasedTxHistoryDB()
+	for _, dbInfo := range dbInfos {
+		match, err := wallet.addressesMatchAccounts(ctx, dbInfo.Addresses)
```
Comment (buck54321) [inline client/asset/dcr/dcr.go:986]: Why do we need different DBs for different account configurations? It's not a new wallet.

Comment (martonp) [reply client/asset/dcr/dcr.go:986]: You need all the accounts to be the same to have the same tx history, not just one. If they change the mixed account, but not the unmixed account, we cannot show the original history.

File: client/asset/dcr/dcr.go [hunk 3]
```diff
@@ -4818,18 +4825,37 @@ func (dcr *ExchangeWallet) shutdown() {
...
+		}
+		dcr.previouslySynced.Store(synced)
+
+		if !previouslySynced && synced {
+			dcr.tipMtx.RLock()
+			tip := dcr.currentTip
+			dcr.tipMtx.RUnlock()
+
+			dcr.syncTxHistory(dcr.ctx, uint64(tip.height))
+		}
```
Comment (buck54321) [inline client/asset/dcr/dcr.go:4843]: A little bit cleaner, I think

```Go
if wasSynced := dcr.previouslySynced.Swap(synced); synced && !wasSynced {
	dcr.tipMtx.RLock()
	tip := dcr.currentTip
	dcr.tipMtx.RUnlock()

	dcr.syncTxHistory(dcr.ctx, uint64(tip.height))
}
```

File: client/webserver/site/src/js/wallets.ts [hunk 4]
```diff
@@ -971,6 +974,7 @@ export default class WalletsPage extends BasePage {
     for (const b of assetSelect.children) b.classList.remove('selected')
     this.assetButtons[assetID].bttn.classList.add('selected')
     this.selectedAssetID = assetID
+    this.page.hideMixTxsCheckbox.checked = true
```
Comment (buck54321) [inline client/webserver/site/src/js/wallets.ts:977]: This should be remembered, but I'm not gonna push it now.

File: client/webserver/site/src/js/wallets.ts [hunk 5]
```diff
@@ -1864,19 +1871,54 @@ export default class WalletsPage extends BasePage {
...
+  async getTxHistory (assetID: number, hideMixTxs: boolean, after?: string) : Promise<TxHistoryResult> {
+    let numToFetch = 10
+    if (hideMixTxs) numToFetch = 15
+
+    const res : TxHistoryResult = { txs: [], lastTx: false }
+    let ref = after
+
+    for (let i = 0; i < 40; i++) {
+      const currRes = await app().txHistory(assetID, numToFetch, ref)
+      if (currRes.txs.length > 0) {
```
Comment (buck54321) [inline client/webserver/site/src/js/wallets.ts:1883]: This is kind of a hack. We should be passing a filter all the way in to the wallet.

**Review by buck54321 — APPROVED** [review]

**Pull request merged**